ee/desktop/notify/notify_linux.go

```diff
@@ -4,38 +4,141 @@
 package notify
 
 import (
+	"context"
 	"fmt"
 	"os/exec"
+	"sync"
+	"time"
 
+	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 	"github.com/godbus/dbus/v5"
 )
 
-func (d *DesktopNotifier) SendNotification(title, body string) error {
-	if err := d.sendNotificationViaDbus(title, body); err == nil {
+type dbusNotifier struct {
+	iconFilepath        string
+	logger              log.Logger
+	conn                *dbus.Conn
+	signal              chan *dbus.Signal
+	interrupt           chan struct{}
+	sentNotificationIds map[uint32]bool
+	lock                sync.RWMutex
+}
+
+const (
+	notificationServiceObj       = "/org/freedesktop/Notifications"
+	notificationServiceInterface = "org.freedesktop.Notifications"
+	signalActionInvoked          = "org.freedesktop.Notifications.ActionInvoked"
+)
+
+func NewDesktopNotifier(logger log.Logger, iconFilepath string) *dbusNotifier {
+	conn, err := dbus.ConnectSessionBus()
+	if err != nil {
+		level.Warn(logger).Log("msg", "couldn't connect to dbus to start notifier listener, proceeding without it", "err", err)
+	}
+
+	return &dbusNotifier{
+		iconFilepath:        iconFilepath,
+		logger:              log.With(logger, "component", "desktop_notifier"),
+		conn:                conn,
+		signal:              make(chan *dbus.Signal),
+		interrupt:           make(chan struct{}),
+		sentNotificationIds: make(map[uint32]bool),
+		lock:                sync.RWMutex{},
+	}
+}
+
+func (d *dbusNotifier) Listen() error {
+	if d.conn != nil {
+		if err := d.conn.AddMatchSignal(
+			dbus.WithMatchObjectPath(notificationServiceObj),
+			dbus.WithMatchInterface(notificationServiceInterface),
+		); err != nil {
+			level.Error(d.logger).Log("msg", "couldn't add match signal", "err", err)
+			return fmt.Errorf("couldn't register to listen to signals in dbus: %w", err)
+		}
+		d.conn.Signal(d.signal)
+	} else {
+		level.Warn(d.logger).Log("msg", "cannot set up DBUS listener -- no connection to session bus")
+	}
+
+	for {
+		select {
+		case signal := <-d.signal:
+			if signal == nil || signal.Name != signalActionInvoked {
+				continue
+			}
+
+			// Confirm that this is a Kolide-originated notification by checking for known notification IDs
```

Comment:
Long term, is this the sort of thing we should have strong authentication for? Should we sign them, for example?

Reply:
At the moment, it's possible but very unlikely that we'd act on a notification that we didn't originate, since we don't clear out these notification IDs from `d.sentNotificationIds` after acting on them. I don't think another application would be allowed to create an `ActionInvoked` signal with a previously-used ID until the ID resets after hitting the max of uint32. So I think in the future possibly worth adding strong authentication, and definitely worth removing a notification ID from `d.sentNotificationIds` after we act on it.

For how we could accomplish this: the only info we get from dbus that is usable (that I've found) is the notification ID assigned by dbus and the URL itself (i.e. the `action_key`) -- so we could possibly encrypt the URL, and then only act on URLs that we can decrypt.